### ISSUE TITLE
Question about reproducing MetaWorld MT50 results

### ISSUE THREAD

**Issue by fengzichao** [MINT-SJTU/Evo-1#34]
Hi, thank you for releasing the code and checkpoints for this great work!

We are trying to reproduce the MetaWorld MT50 results reported in the paper using the released Evo-1 MetaWorld checkpoint. We followed the public evaluation setting as closely as possible:

- Benchmark: MetaWorld MT50
- Episodes per task: 10
- Episode horizon: 400
- Action execution horizon: 15
- Camera: corner2
- Image size: 448 x 448
- Selector: original single-candidate inference, i.e. K=1 / first action chunk
- Seeds we tested: 4042, 4043, 4044, 4045

However, our reproduced K=1 baseline is consistently lower than the paper result. The paper reports an average success rate of around 80.6 on MetaWorld, while our results are:

| Seed | Avg success rate |
|---|---:|
| 4042 | 0.772 |
| 4043 | 0.759 |
| 4044 | 0.758 |
| 4045 | 0.744 |
| Mean | 0.758 |

We are wondering whether we may have missed some details in the evaluation setup.  Could you please help us？
We really appreciate your help. Thanks again for the great work!

**Comment by jywu511** [comment]
The same problem, and I get the success rate 76.2%.  I use the seed 4042.
              reproduced   paper   delta
  Easy          90.0    89.2    +0.8
  Medium        70.0    76.8    -6.8
  Hard          76.7    77.2    -0.5
  Very Hard     68.0    79.2    -11.2
  Avg           76.2    80.6    -4.4

**Comment by DorayakiLin** [comment]
Hi,

Thank you for your interest in our work and for the detailed reproduction setup.

We checked this issue on our side and found that this kind of gap can indeed happen on some machines. In our internal tests, evaluation on RTX 4090D and RTX 4060 stays within the normal fluctuation range of around 2% compared with the reported result.

So it is possible that the difference you observed is related to hardware or environment-specific factors, rather than the evaluation protocol itself.

Thanks again for pointing this out, and we really appreciate your careful reproduction efforts.

**Comment by fengzichao** [comment]
# Unable to reproduce MetaWorld performance with the released dataset and public training recipe

Hi Evo-1 team,

Thanks for releasing the code, dataset, and checkpoints. I tried to reproduce the MetaWorld training result using the released `Evo1_MetaWorld_Dataset` and the public two-stage training recipe, but the checkpoint trained locally performs much worse than the released `Evo1_MetaWorld` checkpoint under the same evaluation protocol.

## Setup

Repository:

```text
Branch: evo1-flash
Commit: d2c53d9a3e0b47b4ea1c8a575bd515f34e0d2706
```

Environment:

```text
Python: 3.10
torch: 2.5.1+cu124
CUDA: 12.4
transformers: 4.39.0
accelerate: 1.13.0
deepspeed: 0.19.0
flash_attn: 2.7.3
GPU: single GPU
```

The dataset appears to be fully downloaded:

```text
Evo1_MetaWorld_Dataset size: 2.3G
No git-lfs pointer files found
data/chunk-000/*.parquet exists
videos/chunk-000/observation.images.image/*.mp4 exists
```

The dataset config I used is:

```yaml
max_action_dim: 24
max_state_dim: 24
max_views: 3
normalization_type: bounds

data_groups:
  metaworld_sawyer:
    Evo1_MetaWorld:
      path: /path/to/Evo1_MetaWorld_Dataset
      view_map:
        image_1: observation.images.image
```

## Training

I followed the public two-stage recipe:

Stage 1:

```text
max_steps=5000
batch_size=16
lr=1e-5
image_size=448
horizon=50
per_action_dim=24
state_dim=24
num_layers=8
finetune_action_head=True
finetune_vlm=False
```

Stage 2:

```text
max_steps=80000
batch_size=16
lr=1e-5
image_size=448
horizon=50
per_action_dim=24
state_dim=24
num_layers=8
finetune_action_head=True
finetune_vlm=True
resume=True
resume_pretrain=True
resume_path=/path/to/stage1/step_5000
```

The stage2 log shows that stage1 was loaded successfully:

```text
Loaded standard checkpoint from /path/to/stage1/step_5000/checkpoint.pt
Resuming pretraining from scratch, resetting step to 0
```

Training completed successfully. The final log includes:

```text
Saved checkpoint to .../stage2/step_80000
Saved checkpoint to .../stage2/step_final
Best checkpoint saved to step_best/ with loss 0.001910
```

The released checkpoint config and the locally trained checkpoint config are mostly consistent in the main training hyperparameters:

```text
vlm_name=OpenGVLab/InternVL3-1B
action_head=flowmatching
use_augmentation=True
lr=1e-5
batch_size=16
max_steps=80000
warmup_steps=1000
grad_clip_norm=1.0
weight_decay=0.001
finetune_vlm=True
finetune_action_head=True
per_action_dim=24
state_dim=24
horizon=50
num_layers=8
dropout=0.2
```

One difference is that the released checkpoint config uses:

```text
resume_path=/path/to/stage1/step_20000
```

while the README-style training recipe uses stage1 `step_5000`.

## Evaluation

I evaluated both checkpoints with the same MetaWorld server/client pipeline. Video/window/image saving was disabled for speed, but the evaluation protocol was otherwise unchanged:

```text
Target difficulty: all
Episodes per task: 10
Episode horizon: 400
HORIZON: 15
Seed: 4042
```

For the locally trained checkpoint, since training saves `checkpoint.pt` with a `model_state_dict`, while the released checkpoint uses `mp_rank_00_model_states.pt` with a `module` key, I converted it before evaluation:

```python
import torch
from pathlib import Path

ckpt_dir = Path("/path/to/stage2/step_80000")
src = ckpt_dir / "checkpoint.pt"
dst = ckpt_dir / "mp_rank_00_model_states.pt"

obj = torch.load(src, map_location="cpu")
state = obj["model_state_dict"]
torch.save({"module": state}, dst)
```

The server then loads the converted checkpoint successfully.

## Results

Under the same evaluation protocol:

```text
Locally trained checkpoint:
easy      : 0.554
medium    : 0.227
hard      : 0.200
very_hard : 0.260
overall   : 0.310

Released Evo1_MetaWorld checkpoint:
easy      : 0.896
medium    : 0.718
hard      : 0.733
very_hard : 0.760
overall   : 0.777
```

Some representative task-level comparisons:

```text
button-press-topdown-v3:      trained 0.500 vs released 1.000
dial-turn-v3:                 trained 0.100 vs released 1.000
faucet-open-v3:               trained 0.100 vs released 1.000
handle-pull-v3:               trained 0.000 vs released 0.500
pick-place-v3:                trained 0.000 vs released 0.500
shelf-place-v3:               trained 0.000 vs released 0.300
```

Since the released checkpoint achieves a reasonable score under the same evaluation setup, the evaluation pipeline seems to be working. However, the public training recipe does not reproduce the released checkpoint performance in my setup.

## Questions

Could you please clarify:

1. Is the public two-stage training recipe sufficient to reproduce the released `Evo1_MetaWorld` checkpoint performance?
2. Was the released `Evo1_MetaWorld` checkpoint trained on `evo1-flash`, `main`, or another exact commit?
3. Should MetaWorld stage1 use `max_steps=5000` as shown in the public recipe, or `step_20000` as suggested by the released checkpoint config?
4. Is the dataset config above correct for `Evo1_MetaWorld_Dataset`, especially the `data_groups`, `view_map`, and camera mapping?
5. Are there any additional image preprocessing details for MetaWorld training, such as camera name, rotation, center crop, resize, or normalization, that should match the evaluation client?
6. Is converting `checkpoint.pt["model_state_dict"]` to `{"module": state_dict}` the correct way to evaluate locally trained checkpoints with `Evo1_server.py`, or is there a recommended evaluation path for training checkpoints?

Any guidance on the exact training/evaluation recipe needed to reproduce the released MetaWorld checkpoint performance would be greatly appreciated.

**Comment by DorayakiLin** [comment]
Hi, thanks for the detailed report.

The released dataset, code, and training recipe are sufficient to reproduce the released MetaWorld checkpoint. The main difference here is the effective training setup. Our released MetaWorld checkpoint was trained with 8 GPUs, with batch size 16 per GPU, so the effective global batch size is 8 × 16 = 128.

In your setup, you used a single GPU with batch size 16. Although the listed `batch_size` and `max_steps` look similar, the effective number of training samples seen by the model is much smaller. Therefore, the model is likely under-trained in terms of epochs/data exposure, which should be the main reason for the performance gap.

For MetaWorld, Stage 1 should use 20,000 steps. The released checkpoint was initialized from the Stage 1 checkpoint at `step_20000`.

For the dataset config, image preprocessing, and checkpoint conversion, no additional modifications are needed. Please follow the released code and configs directly. The default camera mapping, preprocessing, and evaluation pipeline in the open-source code are the ones we used.

So, to better match the released checkpoint, we recommend using the same 8-GPU setup with batch size 16 per GPU. If training on a single GPU, you may need to increase the total training steps accordingly, since the effective global batch size and number of seen samples are much smaller.

**Comment by fengzichao** [comment]
Thanks for the clarification!
